### ISSUE TITLE
Cannot bypass getters with TypeScript

### ISSUE THREAD

**Issue by nterray** [posva/pinia#1605]
### Reproduction

https://stackblitz.com/edit/github-xe2cxw?file=src%2Fstores%2Fcounter.test.ts

### Steps to reproduce the bug

Given the following code:
```typescript
import { defineStore } from 'pinia';
import { createTestingPinia } from '@pinia/testing';

const useStore = defineStore('lol', {
  state: () => ({ n: 0 }),
  getters: {
    double: (state) => state.n * 2,
  },
});
const pinia = createTestingPinia();
const store = useStore(pinia);

store.n++;
store.double = 3;
```

### Expected behavior

TypeScript should not complain.

### Actual behavior

TypeScript raises the following error:
```
TS2540: Cannot assign to 'double' because it is a read-only property.
```

### Additional information

This is to be able to mock getters in tests
https://pinia.vuejs.org/cookbook/testing.html#mocking-getters

**Comment by LinusBorg** [comment]
Why would you assign to a getter? It's a getter, it has not setter.

TS is warning you correctly here. if you want to do this for some exotic reason: `//@ts-expect-error`

**Comment by nterray** [comment]
Sorry I forgot to mention that it was to be able to mock getters in tests
https://pinia.vuejs.org/cookbook/testing.html#mocking-getters

**Comment by nterray** [comment]
By the way TypeScript complains about this error for [Pinia's test file](https://github.com/vuejs/pinia/blob/pinia%402.0.21/packages/testing/src/testing.spec.ts) :
```
➜ ./node_modules/.bin/tsc packages/testing/src/testing.spec.ts 
[…]
packages/testing/src/testing.spec.ts:217:11 - error TS2540: Cannot assign to 'double' because it is a read-only property.

217     store.double = 3
              ~~~~~~

```

**Comment by posva** [comment]
I think we made the getters read only at some point because its easier to spot bugs.
A possible solution for this is to have some kind of wrapper when calling useStore but still not ideal. At the moment, a ts-expect-error comment in tests will do but maybe we can come up with a more ergonomic solution

**Comment by Hatzen** [comment]
Can we have an update for the docs at least to mention this issue?

**Comment by sarahsporck** [comment]
I adjusted the mockedStore implementation in the docs & made the getters writable in typescript. This seems to work quite fine.

```typescript
type Writeable<T> = { -readonly [P in keyof T]: T[P] };

function mockedStore<TStoreDef extends () => unknown>(
  useStore: TStoreDef,
): TStoreDef extends StoreDefinition<infer Id, infer State, infer Getters, infer Actions>
  ?
      | Exclude<
          Store<
            Id,
            State,
            Getters,
            {
              [K in keyof Actions]: Actions[K] extends (...args: infer Args) => infer ReturnT
                ? // 👇 depends on your testing framework
                  Mock<Args, ReturnT>
                : Actions[K];
            }
          >,
          _StoreWithGetters<Getters>
        >
      | Writeable<_StoreWithGetters<Getters>>
  : ReturnType<TStoreDef> {
  return useStore() as any;
}

const mockedCounterStore = mockStore(useCounterStore);
```

**Comment by posva** [comment]
I forgot to update the docs with the Getters, it's fixed now! See 7218691

**Comment by rdoepke** [comment]
We finally could get rid of hundreds of `@ts-expect-error https://github.com/vuejs/pinia/issues/1605` as well.

However, cookbook solution doesn't work well with setup style stores and build in methods, we make use of as well. Gemini came up with a really nice solution for this - maybe it helps. Here is the variant for vitest - it doesn't even use "as any":

```typescript
import { Mock } from 'vitest';

export type MockedStore<TStoreDef extends () => unknown> = {
  -readonly [K in keyof ReturnType<TStoreDef>]: ReturnType<TStoreDef>[K] extends (...args: any[]) => any
    ? K extends `$${string}`
      ? ReturnType<TStoreDef>[K]
      : Mock<ReturnType<TStoreDef>[K]>
    : ReturnType<TStoreDef>[K];
};

export function mockedStore<TStoreDef extends () => unknown>(useStore: TStoreDef): MockedStore<TStoreDef> {
  return useStore() as MockedStore<TStoreDef>;
}
```
